pennylane/templates/constructors/broadcast_double.py

```diff
@@ -0,0 +1,237 @@
+# Copyright 2018-2019 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+r"""
+Contains the ``broadcast`` template constructor.
+"""
+# pylint: disable-msg=too-many-branches,too-many-arguments,protected-access
+from collections import Iterable
+
+from pennylane.templates.decorator import template
+from pennylane.templates.utils import _check_wires, _check_type, _get_shape
+
+
+@template
+def broadcast_double(block, wires, parameters=None, even=True, kwargs={}):
+    r"""Applies a two-qubit unitary to pairs of wires.
+
+    If ``even`` is ``True``, start with the wire pair ``[0, 1]``, else with the pair ``[1, 2]``.
+
+    .. figure:: ../../_static/templates/constructors/broadcast_double.png
+        :align: center
+        :width: 40%
+        :target: javascript:void(0);
+
+    Each unitary ``block`` applied to a pair of wires may depend on parameters. These are passed as a list
+    by the ``parameters`` argument. For :math:`M` wires and ``even=True``, parameters for
+    :math:`\lfloor \frac{M}{2} \rfloor` unitaries must be supplied, while for ``even=False``
+    there are :math:`\lfloor \frac{M-1}{2} \rfloor` unitaries.
+
+    Typically, ``block`` is either a two-qubit quantum operation (such as :meth:`~.pennylane.ops.CNOT`), or a
+    user-supplied template acting on two qubits. For more details, see *UsageDetails* below.
+
+    Args:
+        block (function): quantum gate or template
+        parameters (iterable or None): sequence of parameters for each gate applied
+        even (boolean): if ``True``, pairs are formed starting with the first wire index, else with
```

Comment:
I'm most probably missing the point here! Since the user is expected to pass the list of wires anyways, could they just not be made to omit the first one if they would like to go with the `even==False` case?

Or if the idea is that most of the times they would go with `range(len(dev.wires))`, we could also document an example with having the counterpart `range(1, len(dev.wires))`.